components/script/dom/element.rs

```diff
@@ -998,14 +1001,71 @@ impl Element {
         }
     }
 
+    /// Check if an element is in the containing block chain for scrollIntoView
+    /// Only elements in the containing block chain should be scrolled by scrollIntoView
+    fn is_in_containing_block_chain(&self, ancestor: &Node) -> bool {
+        if ancestor.is::<Document>() {
+            true // Document viewport is always in the containing block chain
+        } else if ancestor.is::<Element>() {
+            let ancestor = ancestor.downcast::<Element>().unwrap();
+            // Walk up the DOM from self to find if ancestor is in the containing block chain
+            let mut current_node = Some(DomRoot::from_ref(self.upcast::<Node>()));
+
+            while let Some(node) = current_node {
+                if let Some(element) = node.downcast::<Element>() {
+                    // If we reached the ancestor, it's in the containing block chain
+                    if std::ptr::eq(element, ancestor) {
+                        return true;
+                    }
+
+                    // Check if this element establishes a new containing block
+                    if let Some(style) = element.style() {
+                        let position = style.get_box().clone_position();
+
+                        // If this element is absolutely or fixed positioned,
+                        // it establishes a new containing block - stop here
+                        if position.is_absolutely_positioned() && !std::ptr::eq(element, self) {
+                            return false;
+                        }
+                    }
+                }
+
+                // Move to parent
+                current_node = node.GetParentElement().map(DomRoot::upcast);
+            }
+
+            false
+        } else {
+            false
+        }
+    }
```

Comment:
Please use an early return at the top of the function instead of the large nested block.